### PR TITLE
Raise the bar to ISO C99 (drop support for C89)

### PR DESCRIPTION
**Pull request by alejandro-colomar**
<https://github.com/shadow-maint/shadow/issues/600>

### CODE REVIEW

**Comment by alejandro-colomar** [comment]
I think this is all for C99.

After this is merged, I'll suggest raising the bar even further, to POSIX.1-2001.

**Comment by alejandro-colomar** [comment]
Rebased to master.

File: lib/defines.h [hunk 1]
```diff
@@ -250,7 +250,6 @@ static inline void memzero(void *ptr, size_t size)
 #ifdef sun			/* hacks for compiling on SunOS */
 # ifndef SOLARIS
 extern char *strdup ();
-extern char *strerror ();
```
Comment (ikerexxe) [inline lib/defines.h:278]: `configure.ac` also contains a reference to `strerror()` which I believe can be removed.

**Review by ikerexxe — COMMENTED** [review]
Apart from a little comment inline, I'd also like to ask you to add the reference to the issue fixed in the commit messages. Something like:
```
Resolves: https://github.com/shadow-maint/shadow/issues/600
```

File: lib/defines.h (hunk 1 above)
Comment (alejandro-colomar) [reply lib/defines.h:278]: Thanks!

**Comment by alejandro-colomar** [comment]
> Apart from a little comment inline, I'd also like to ask you to add the reference to the issue fixed in the commit messages. Something like:
> 
> ```
> Resolves: https://github.com/shadow-maint/shadow/issues/600
> ```

Yep, I think that would be `Closes:`.  Although since the issue really covers more than just C99, I think I'll just use `Link:`.  Are you happy with it?

**Comment by alejandro-colomar** [comment]
Changes:

- Remove reference to strerror(3) in configure.ac.  [Iker]

**Comment by alejandro-colomar** [comment]
Okay, now it should be done.

Changes:

- Also remove checks for 3 other functions required by C99 (and in some cases even C89).

**Comment by ikerexxe** [comment]
> Yep, I think that would be conventionally `Closes:`. Although since the issue really covers more than just C99, I think I'll just use `Link:`. Are you happy with it?

As long as you reference the issue I'm fine.

LGTM. Thanks for the patches!

**Comment by alejandro-colomar** [comment]
:-)

**Comment by alejandro-colomar** [comment]
I'm happy to help.  And BTW, I learnt a lot with these patches.  Just FYI, I started rewriting some of the Linux man-pages about strings, and plan a new `string(7)` manual page that details all known string handling functions (and some more that I've only seen customly written in high-performance projects such as nginx), and add details when you should use each of them.

As a starter, I already killed `strncpy(3)`, in the hope that nobody ever uses it again ;)
<https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/tree/man3/strncpy.3#n26>

**Comment by alejandro-colomar** [comment]
Great! :-)

**Pull request merged**